File: symmetric-tree/bfs-solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    var result = true;
    if(!root) return result;
    const queue = [root];
    const nodesValByLevel = {};
    var level = 0;
    while(queue.length) {
        const size = queue.length;
        level += 1;
        for (var i = 0; i < size; i++) {
            var node = queue.shift();
            var val = "";
            if (!node) {
                 val = "#"
            } else {
                val = node.val;
            }
            if (nodesValByLevel[level] === undefined) {
                nodesValByLevel[level] = [val];
            } else {
                nodesValByLevel[level].push(val);
            }
            if (!node) continue;
            queue.push(node.left, node.right);
        }
    }
    for (var level in nodesValByLevel) {
        const values = nodesValByLevel[level];
        const size = values.length;
        const left = values.slice(0, size/2);
        const right = values.slice(size/2).reverse();
        if (values.length > 1 && left.join("") !== right.join("")) {
            result = false;
            break;
        }
    }
    return result;
};

File: binary-tree-level-order-traversal/bfs-solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    var result [][]int
    if root == nil {
        return result
    }
    
    var level int = 1
    var results map[int][]int = make(map[int][]int, 0)
    
    dfs(root, results, level)
    
    var levels int = len(results)
    for i := 1; i <= levels; i++ {
        values := results[i]
        result = append(result, values)
    }
    
    return result
}

func dfs(node *TreeNode, results map[int][]int, level int) {
    if node == nil {
        return
    }
    
    if _, found := results[level]; found {
        results[level] = append(results[level], node.Val)
    } else {
        results[level] = []int{node.Val}
    }
    
    dfs(node.Left, results, level + 1)
    dfs(node.Right, results, level + 1)
}